src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataBlockMigrationService.cs

```diff
@@ -1,40 +1,456 @@
+        {
+            DateTime createdDate;
+
+            if (dataBlock.Created != null)
+            {
+                createdDate = dataBlock.Created.Value;
+            }
+            else
+            {
+                var owningRelease = (await _contentDbContext
+                    .ReleaseContentBlocks
+                    .Include(rcb => rcb.Release)
+                    .SingleAsync(rcb => rcb.ContentBlockId == dataBlock.Id))
+                    .Release;
+
+                createdDate = owningRelease.Created;
+            }
+
+            return createdDate;
+        }
+
+        private 
+            List<(
+                GeographicLevel geographicLevel, 
+                DateTime start, 
+                DateTime end,
+                BoundaryLevel boundaryLevel
+            )>
+            GetBoundaryLevelsWithDateRanges(List<BoundaryLevel> boundaryLevels)
+        {
+            var groupedByGeographicLevel = boundaryLevels
+                .GroupBy(boundary => boundary.Level)
+                .ToDictionary(
+                    group => group.Key,
+                    group => group.OrderBy(boundary => boundary.Created).ToList());
+
+            return groupedByGeographicLevel
+                .SelectMany(levelAndBoundaries =>
+                {
+                    var (level, boundaries) = levelAndBoundaries;
+                    return boundaries
+                        .Select(boundary =>
+                        {
+                            var indexOfBoundary = boundaries.IndexOf(boundary);
+                            var previousBoundary = indexOfBoundary > 0 ? boundaries[indexOfBoundary - 1] : null;
+                            var nextBoundary = indexOfBoundary < boundaries.Count - 1 ? boundaries[indexOfBoundary + 1] : null;
+                            return (
+                                level,
+                                previousBoundary == null ? boundary.Created.AddYears(-10) : boundary.Created,
+                                nextBoundary?.Created.AddMilliseconds(-1) ?? boundary.Created.AddYears(10),
+                                boundary
+                            );
+                        });
+                })
+                .ToList();
+        }
+
+        private BoundaryLevel GetLatestBoundaryLevelAtCreationTime(
+            DateTime dataBlockCreated,
+            GeographicLevel geographicLevel, 
+            List<BoundaryLevel> boundaryLevels)
+        {
+            var boundaryLevelsWithDateRanges = GetBoundaryLevelsWithDateRanges(boundaryLevels);
+            return boundaryLevelsWithDateRanges
+                .Single(boundary =>
+                {
+                    var (level, start, end, _) = boundary;
+
+                    return level == geographicLevel &&
+                           dataBlockCreated.CompareTo(start) >= 0 &&
+                           dataBlockCreated.CompareTo(end) <= 0;
+                })
```

Comment:
As I understand it, this works because `boundaryLevelsWithDateRanges` is ordered? So it selects the earliest created boundary level that the data block's creation date is within its start/end?

Reply:
Yup that's correct.